examples/gan/_gan_mnist_pl/pl.py

```diff
@@ -0,0 +1,262 @@
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(512, 256),
+            nn.LeakyReLU(0.2, inplace=True),
+            nn.Linear(256, 1),
+            nn.Sigmoid(),
+        )
+
+    def forward(self, img):
+        img_flat = img.view(img.size(0), -1)
+        validity = self.model(img_flat)
+
+        return validity
+
+
+"""### C. GAN
+#### A couple of cool features to check out in this example...
+
+  - We use `some_tensor.type_as(another_tensor)` to make sure we initialize new tensors on the right device (i.e. GPU, CPU).
+    - Lightning will put your dataloader data on the right device automatically
+    - In this example, we pull from latent dim on the fly, so we need to dynamically add tensors to the right device.
+    - `type_as` is the way we recommend to do this.
+  - This example shows how to use multiple dataloaders in your `LightningModule`.
+"""
+class GAN(pl.LightningModule):
+
+    def __init__(
+        self,
+        channels,
+        width,
+        height,
+        latent_dim: int = 100,
+        lr: float = 0.0002,
+        b1: float = 0.5,
+        b2: float = 0.999,
+        batch_size: int = 64,
+        **kwargs
+    ):
+        super().__init__()
+        self.save_hyperparameters()
+
+        # networks
+        data_shape = (channels, width, height)
+        self.generator = Generator(latent_dim=self.hparams.latent_dim, img_shape=data_shape)
+        self.discriminator = Discriminator(img_shape=data_shape)
+
+        self.validation_z = torch.randn(8, self.hparams.latent_dim)
+
+        self.example_input_array = torch.zeros(2, self.hparams.latent_dim)
+
+    def forward(self, z):
+        return self.generator(z)
+
+    def adversarial_loss(self, y_hat, y):
+        return F.binary_cross_entropy(y_hat, y)
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
+        imgs, _ = batch
+
+        # sample noise
+        z = torch.randn(imgs.shape[0], self.hparams.latent_dim)
+        z = z.type_as(imgs)
+
+        # train generator
+        if optimizer_idx == 0:
+
+            # generate images
+            self.generated_imgs = self(z)
+
+            # CHANGE remove logger
+            # log sampled images
+            sample_imgs = self.generated_imgs[:6]
+            grid = torchvision.utils.make_grid(sample_imgs)
+            # self.logger.experiment.add_image('generated_images', grid, 0)
+
+            # ground truth result (ie: all fake)
+            # put on GPU because we created this tensor inside training_loop
+            valid = torch.ones(imgs.size(0), 1)
+            valid = valid.type_as(imgs)
+
+            # adversarial loss is binary cross-entropy
+            g_loss = self.adversarial_loss(self.discriminator(self(z)), valid)
+            tqdm_dict = {'g_loss': g_loss}
+            output = OrderedDict({
+                'loss': g_loss,
+                # 'progress_bar': tqdm_dict, #  FIXME unserializable object torch.Tensor
```

Comment:
```
 File "/run/determined/pythonuserbase/lib/python3.6/site-packages/determined/util.py", line 150, in json_serializer
<[]>
raise TypeError("Unserializable object {} of type {}".format(obj, type(obj)))
```

Reply:
I found in pytorch trial we don't support returning nested dictionaries in metrics yet.